File: C#/ShoppingCartApp/ShoppingCart.cs
using System;
using System.Collections.Generic;

namespace ShoppingCartApp
{
    public class ShoppingCart<T> where T : notnull
    {
        private Dictionary<T, int> items = new Dictionary<T, int>();

        public void AddItem(T item, int quantity)
        {
            if (items.ContainsKey(item))
            {
                items[item] += quantity;
            }
            else
            {
                items[item] = quantity;
            }
        }

        public void RemoveItem(T item)
        {
            if (items.ContainsKey(item))
            {
                items.Remove(item);
            }
            else
            {
                throw new InvalidOperationException("Item not found in the cart.");
            }
        }

        public int GetTotalItems()
        {
            int total = 0;
            foreach (var quantity in items.Values)
            {
                total += quantity;
            }
            return total;
        }

        public void PrintCart()
        {
            foreach (var item in items)
            {
                Console.WriteLine($"{item.Key}: {item.Value}");
            }
        }
    }
}

File: C++/ShoppingCartApp/shopping_cart.cpp
#include "shopping_cart.h"

template <typename T>
void ShoppingCart<T>::AddItem(T item, int quantity) {
    auto it = std::find_if(items.begin(), items.end(), [&item](const std::pair<T, int>& p) {
        return p.first == item;
    });
    if (it != items.end()) {
        it->second += quantity;
    } else {
        items.push_back(std::make_pair(item, quantity));
    }
}

template <typename T>
void ShoppingCart<T>::RemoveItem(T item) {
    items.erase(std::remove_if(items.begin(), items.end(), [&item](const std::pair<T, int>& p) {
        return p.first == item;
    }), items.end());
}

template <typename T>
int ShoppingCart<T>::GetTotalItems() const {
    int total = 0;
    for (const auto& item : items) {
        total += item.second;
    }
    return total;
}

template <typename T>
void ShoppingCart<T>::PrintCart() const {
    for (const auto& item : items) {
        std::cout << "Item: " << item.first << ", Quantity: " << item.second << std::endl;
    }
}

template class ShoppingCart<int>;
template class ShoppingCart<std::string>;